File: c++/binary_search/findProduct.cpp
#include <iostream>
#include <vector>
using namespace std;
vector<int> wanted;
int N , M;
vector<int> products;

bool findProduct(int start, int end ,int wantedProduct){
    int mid = (start +end )/ 2;

    if(start > end) return false; //>=?

    if(products[mid] == wantedProduct)
        return true;
    else if(products[mid] > wantedProduct)
        return findProduct(start, mid - 1, wantedProduct);

    return findProduct(mid + 1, end, wantedProduct);
}

int main(){
    cin >> N;

    for(int i = 0 ;i<N ; i++){
        int tmp ;
        cin >> tmp;
        products.push_back(tmp);
    }
    cin >> M;
    for(int i = 0 ;i<M ; i++){
        int tmp ;
        cin >> tmp;
        wanted.push_back(tmp);
    }

    for(int i = 0 ;i<M ; i++){
        if(findProduct(0,N-1,wanted[i]))
            cout << "YES ";

        else
            cout << "NO ";
    }





}

/*
 5
 8 3 7 9 2
 3
 5 7 9
 */

File: python/binary_search/findProduct.py

import sys;
n = int(input())

products = list(map(int,sys.stdin.readline().rstrip().split()))
products.sort()

m = int(input())

wanted = list(map(int,sys.stdin.readline().rstrip().split()))



def findProduct(start, end , product):
    if start > end :
        return False
    mid = (start + end )//2 

    if products[mid] == product :
        return True
    elif products[mid] > product :
        return findProduct(start, mid-1 , product)
    else :
        return findProduct(mid+1, end , product)



for i in range(len(wanted)):
    if findProduct(0,n-1,wanted[i]) == True:
        print("Yes ",end=' ')
    else:
        print("No ",end=' ')
